### PR TITLE
doc: Separate repository for the gui

### PR DESCRIPTION
**Pull request by maflcko**
## 🥅 Goals

Splitting up the GUI (and splitting out modules in general) has been brought up often in recent years. Now that the GUI is primarily connected through (internal) interfaces with the node, it seems an appropriate time to revive this discussion.

Before looking for solutions, we should define a set of goals that we want to achieve. I will start with some ideas to get started and I hope that others will chime in to share and prioritize their goals.

### Separate issue and patch management

It is currently not possible to subscribe to only a subset of modules in Bitcoin Core, or exclude modules from issue and patch notifications. While it is possible to reactively mute conversations in the stream of all ongoing discussions, there is no way to proactively achieve this. Moreover, the list of open issues and pull request will always include GUI related ones by default. Only with [filters](https://github.com/bitcoin/bitcoin/pulls?q=is%3Aopen+is%3Apr+-label%3AGUI) it is possible to hide them.

### More focused review and interests

Long term goals of the GUI are partially unclear #17395 . Bitcoin Core developers are generally fluent on the command line. Thus, they might not be interested or motivated to review improvements to the GUI, which might not affect their workflow on the command line at all.  Splitting up the GUI will hopefully attract similar minded people to a project whose primary goal is to build and improve the GUI.

### Maintain high quality assurance

The quality of the GUI (and even more importantly Bitcoin Core in general) must not degrade. This means that code review itself is not negatively affected by splitting the GUI, but also the integration of the GUI into the rest of Bitcoin Core. One issue could arise when arbitrary version-combinations are allowed. We are struggling hard to test against all supported versions of Boost. Making the GUI version another dimension is going to make testing impossible.

### The GUI *is* Bitcoin Core

When a user downloads Bitcoin Core from our website (or another package manager) they expect the GUI to be included. This should not change (at least not as a result of splitting up the GUI into another project).

Similarly, when building Bitcoin Core, the gui should still be built when `--with-gui` is specified.

## 🌳 Proposed solution: Monotree

TLDR. Everything stays the same, the development process for the GUI changes slightly.

Long version:
* An exact mirror of the master development branch is hosted at `bitcoin-core/gui`. The new repository is used to track gui-only issues and pull requests. Global changes that happen to touch gui code still go to the *main* repo.
* All pull requests will be merged into `bitcoin/bitcoin`.
* Decision making process and maintainers will be identical for both repos.

### Disadvantages

* Review activity might decrease?

* It doesn't go far enough. bitcoin/bitcoin#3440 is proposing a modularized Bitcoin Core. The GUI could be an "add-on", connected over RPC or capnproto (bitcoin/bitcoin#10102). Thus, the gui could even be hosted as a subtree or completely separate project.

### Advantages

* Review activity might increase? It is impossible to predict the future, but for example the `libsecp256k1` subtree has a lot of domain specific experts, maintainers and reviewers. I think longer term it makes sense to at least try this route for the gui as well.

* A smaller step is easier to undo when it turns out to come with any unforeseen downsides.

* No substantial changes to the decision making progress.

* Nothing changes in how developers set up their dev environment or how users build from the source. Also, the release binaries and process will stay exactly the same. No version drift. Finally, code sharing between the GUI and Bitcoin Core is not made any harder.

* The organizational side. There are 72 open issues (~14%) and 61 open PRs (~16%) with the GUI label. If moved to its own repo, non-GUI developers wouldn't have to be distracted with GUI-only issues and PRs and GUI enhancements. GUI developers have their own repo to focus on GUI development exclusively.

### Implementation (outstanding TODOs)

* Adjust maintainer merge script https://github.com/bitcoin-core/bitcoin-maintainer-tools/pull/57
* Create bitcoin-core/gui repository (empty or with master branch only)
* Assign all existing bitcoin core maintainers to the new repo
* Celebrate? :partying_face:

* Long-term: Think how long the grace period is for existing GUI related issues and pull requests. Issues can be transferred with a script after a grace period of some months?

### CODE REVIEW

**Comment by MarcoFalke** [comment]
There are some unrelated doc changes that I will need to split out.

Also, thanks to everyone who provided feedback or ideas that lead to this proposal!

**Comment by jonatack** [comment]
YMMV, but here's an anecdote. I was lead maintainer in a team of four maintainers of a popular open source database search engine from 2013-2016. About a year after I decided to move on, it was decided to do something similar to this proposal for a part of the project. Quite some time was spent to separate it. After that, the split-off part of the codebase saw much less attention and love. Its contributors lost motivation. Two years after the split-off, it was decided to reintegrate it back into the project as before. Much dev time went into these transitions.

**Comment by laanwj** [comment]
Concept ACK. While I agree with @jonatack's point that tracking too many different repositories adds a burden, if it's all just the same people, I think we've been running against against code review bottleneck for this repository for a long time.

The GUI is a suitably separate thing that it'd warrant being maintained separately. It could use some fresh energy (with some different maintainers enthusiastic about it, I'd hope) and I don't think that can happen as long as it's part of this repository.

This doesn't get us there, but it's at least a step.

**Comment by kristapsk** [comment]
Couldn't git submodules be used to separate GUI to separate repository, but still allow to build GUI together with everything else?

File: CONTRIBUTING.md
Comment (flack) [inline CONTRIBUTING.md:78]: So if a PR only adds a new GUI feature, but needs to touch a couple of lines in other areas for that, does it go into the main repo or the gui one? (and would drahtbot be able to detect conflicts if it goes to the gui one?)

Comment (MarcoFalke) [reply CONTRIBUTING.md:78]: I will run DrahtBot on the gui repo when it is created. Given that the GUI speaks with the node over interfaces, there shouldn't be too many inter-repo conflicts.

**Comment by jonasschnelli** [comment]
Thanks for bringing this up.
Concept ACK (in the long run, unsure about the way to proceed).

* I think as long as the GUI is still welded together with the Core part, a submodule-ish approach (like libsecp256k1) makes little sense.

* The hoped benefits are currently not git related but GitHub/Communication related. Therefore we should always question how much actual source code changes are worth implementing to bypass a communication platform issue (not saying this PR does that).

* Unsure if the described mono-tree approach brings a relief in communication. Things like `For global refactoring or other global changes that touche GUI code, as well as
 non-GUI code, the main repository should be used.` makes it even more complex where to do what and how to change it if a PR suddenly needs to touch Core code.

* Also unsure if the good experience we made with domain specific experts working on libsecp256k1 would also happen with the GUI. Libsecp256k1 is useful for a lot of projects (1:n) where the Bitcoin-Core-GUI would only have a specific use case (1:1). Testing the GUI and working on the GUI needs good insights into the Core repository as well (not the case for libsecp256k1).


Splitting of the GUI code and communication is probably undoubtable a great long term achievement. I just think we first need better isolation/seperation for the GUI. A feasible "API contract" how the GUI can asynchronously work with the Core part in a separate process is IMO a precondition to split the GUI into a separate repository. Otherwise I see the risk of actually increase the complexity and work-load on both sides.

**Comment by MarcoFalke** [comment]
Looking at the last 15 or so gui related pull requests, the only one that was not gui-only was one pull request involving cs_main (#19011). For that reason it needed to also touch the node interface code.

I might do a review of all open pull requests that have "qt" or "gui" in the title, but my impression is that most are already gui-only (only changing code in the `src/qt` directory).

**Comment by fanquake** [comment]
> One issue could arise when arbitrary version-combinations are allowed.

Why would the split out GUI require versioning? It's still tightly coupled (build system, dependencies, qt version, minimum requirements wise etc) to the main repo, and the GUI repo would have no scope to change anything significant in that regard. Or do you mean this would be an issue if the GUI actually was to become an entirely separate project?

> Review activity might decrease?
> Review activity might increase?

My guess is decrease. The libsecp256k1 repo is very different. It's critical infrastructure, has contributors that are active across it, bitcoin/bitcoin, and many other projects, and there's far less scope for the type of bike-shedding/personal opinion (my distro does X!) type discussion that would undoubtedly occur in a GUI only-repo. I'd assume that while some of the reviewers that may have worked on GUI PRs, may also head to the alt repo, most may just stay and migrate to different parts of the codebase.

> A smaller step is easier to undo when it turns out to come with any unforeseen downsides.

Do you mean a small step change to the GUI? If we're going to have to start reverting GUI merges because not enough people looked at the other repo before something was merged here, that would not be an improvement to the workflow of this repository or the project in general.

> The organizational side. There are 72 open issues (~14%) and 61 open PRs (~16%)

> Looking at the last 15 or so gui related pull requests, the only one that was not gui-only was one pull request involving cs_main (#19011). For that reason it needed to also touch the node interface code.

Looking at the open GUI tagged PRs: #18927, #18542, #18165, #18027, #17993, #17227, #16883, #16549 are some that touch areas outside the GUI.

> Long-term: Think how long the grace period is for existing GUI related issues and pull requests. Issues can be transferred with a script after a grace period of some months?

I don't think we want a long grace period.  Otherwise we're just living in the worst of both worlds. It also means a longer time for cross-referencing (in issues) to build up between the two repos, and discussions to be(come) fragmented.


Moving the GUI out of this repo is something I've thought about for quite some time, so Concept ACK, even though as mentioned, this would just be a partial step in (hopefully) the right direction.

Honestly I think it's unlikely that I would want to follow the new GUI repo issue/PR discussion (i.e fully subscribed as I am here), or review and merge changes from it. Although that's probably a little dependent on what they end up looking like.

**Comment by MarcoFalke** [comment]
> Or do you mean this would be an issue if the GUI actually was to become an entirely separate project?

Yes, this is only a theoretical concern, if the GUI should ever become a separate software project/git tree.

> If we're going to have to start reverting GUI merges because not enough people looked at the other repo before something was merged here, that would not be an improvement to the workflow of this repository or the project in general.

One of the goals is to maintain high quality assurance. Thus, the proposed solution explicitly does *not* change the decision making process and the maintainers from the status quo.

> I don't think we want a long grace period. Otherwise we're just living in the worst of both worlds. It also means a longer time for cross-referencing (in issues) to build up between the two repos, and discussions to be(come) fragmented.

To clarify, the grace period would only affect "grandfathered" pull requests, not new ones. I would feel uncomfortable closing all gui related pull requests in this repo over night (even the ones that are almost ready for merge). I think we should give each pull request that has already been filed against this repo a chance to make progess and get finalized (merged or closed) before transferring  it to the new repo. Otherwise it will be the exact outcome that you try to avoid. ("discussions to be(come) fragmented")

> Looking at the open GUI tagged PRs: #18927, #18542, #18165, #18027, #17993, #17227, #16883, #16549 are some that touch areas outside the GUI.

* #18927 bilingual_str is an interface to provide translations. Primarily for the GUI, but other tools could also use them. So this interface pull request should be handled in bitcoin/bitcoin, I think.
* #18542 here as well, the only "non-gui" change is interface related. Maybe we could extend the rule from "gui-only means src/qt" to include changes in the interfaces folder, but I'll leave that up for discussion.
* #18165 is a "global refactoring" affecting GUI and RPC code. The goal of this pull is to share pure utility methods between both components. This pull would be correctly filed against bitcoin/bitcoin.
* #18027 The non-gui changes (psbt and wallet changes that affect rpc) should be split out. This is already the policy today, unless I am mistaken?
* #17993 Only interface changes. See #18542 above.
* #17227, #16883 At the very least the build system changes are not strictly GUI related and should stay in this repo.
* #16549 The non-gui changes (psbt and wallet changes that affect rpc) should be split out. This is already the policy today, unless I am mistaken?

**Comment by MarcoFalke** [comment]
> Maybe we could extend the rule from "gui-only means src/qt" to include changes in the interfaces folder, but I'll leave that up for discussion.

Added a TODO to allow for some time to think about this.

**Comment by jb55** [comment]
Concept ACK, the linux-like monotree model makes a lot of sense. I mainly use email to track issues and review code, so having things split up like how lkml does it would be great. Right now it is a bit overwhelming with the number of PRs. Splitting out into separate topics might help with that and improve focus/ for each subsystem. I'm not sure it would make sense to have any other repos like one for network, etc. but one for gui seems reasonable.

**Comment by practicalswift** [comment]
Concept ACK: allowing for increased specialization is good :)

File: CONTRIBUTING.md
Comment (Sjors) [inline CONTRIBUTING.md:79]: 992f5b56195c0bb3d4a12c1fc7f2af7de39d6cdd nit: "Consider making two separate pull requests."

File: CONTRIBUTING.md
Comment (Sjors) [inline CONTRIBUTING.md:82]: If you leave the todo, point it to #19071

File: CONTRIBUTING.md
Comment (Sjors) [inline CONTRIBUTING.md:77]: nit: "touch"

**Comment by Sjors** [comment]
Concept ACK

It's the most light-weight way to move to a new repository. Later on I image we'll use a submodule or some other way to separate the codebase itself, but for now we don't have to commit to the exact method.

Regarding #16549 `UI external signer support (e.g. hardware wallet)`:
> The non-gui changes (psbt and wallet changes that affect rpc) should be split out. This is already the policy today, unless I am mistaken?

Most of that is already separated into #16546. But e.g. 858d9555e27576b6abd618101f5717715839b723 adds a method to `src/interfaces/wallet.h` which isn't needed by the RPC. So that would have to go into another PR in this new scheme, which is fine by me.

That PR requires constant rebasing on the non-GUI changes it needs. With a monotree which tracks master, that's not a problem. But if we end up splitting the project more strongly, then it's important that interface changes are available to the GUI repo quickly.


Can @DrahtBot still work between the two repositories?

**Review by Sjors — APPROVED** [review]
Concept ACK

It's the most light-weight way to move to a new repository. Later on I image we'll use a submodule or some other way to separate the codebase itself, but for now we don't have to commit to the exact method.

Regarding #16549 `UI external signer support (e.g. hardware wallet)`:
> The non-gui changes (psbt and wallet changes that affect rpc) should be split out. This is already the policy today, unless I am mistaken?

Most of that is already separated into #16546. But e.g. 858d9555e27576b6abd618101f5717715839b723 adds a method to `src/interfaces/wallet.h` which isn't needed by the RPC. So that would have to go into another PR in this new scheme, which is fine by me.

That PR requires constant rebasing on the non-GUI changes it needs. With a monotree which tracks master, that's not a problem. But if we end up splitting the project more strongly, then it's important that interface changes are available to the GUI repo quickly.


Can @DrahtBot still work between the two repositories?

File: CONTRIBUTING.md
Comment (MarcoFalke) [reply CONTRIBUTING.md:82]: I am hoping that someone comes up with a guideline here. Otherwise I will flip a coin.

Comment (Sjors) [reply CONTRIBUTING.md:82]: I would stick to defining interfaces in the main repo for now.

**Comment by harding** [comment]
Concept ACK.

**Comment by MarcoFalke** [comment]
There has been some discussion on IRC: http://www.erisian.com.au/bitcoin-core-dev/log-2020-05-28.html#l-406

Touched a lot of points, but the relevant part for the changes in this pull are that `./src/interfaces` is considered node code.

File: CONTRIBUTING.md
Comment (MarcoFalke) [reply CONTRIBUTING.md:78]: Clarified, hopefully.

File: CONTRIBUTING.md
Comment (MarcoFalke) [reply CONTRIBUTING.md]: Thanks, fixed.

**Comment by practicalswift** [comment]
ACK fa2f6ce5d8eef23edcfa38c220d4ac7468acd126: worth testing and easy to revert in the future if we don't achieve the stated goals

**Review by hebasto — APPROVED** [review]
ACK fa2f6ce5d8eef23edcfa38c220d4ac7468acd126, but couldn't verify links to https://github.com/bitcoin-core/gui :)

Let's try it!

**Comment by MarcoFalke** [comment]
(Force pushed a minor reword since the previous review)

File: .github/PULL_REQUEST_TEMPLATE.md
Comment (jonatack) [inline .github/PULL_REQUEST_TEMPLATE.md:7]: s/GUI related/GUI-related/ here and in the other docs in this PR

File: CONTRIBUTING.md
Comment (jonatack) [inline CONTRIBUTING.md:90]: s/gui/GUI/ (either all in uppercase, or all in lowercase)

Comment (jonatack) [reply CONTRIBUTING.md:90]: s/However,/However:/ (or "More specifically:")

File: CONTRIBUTING.md
Comment (jonatack) [inline CONTRIBUTING.md:95]: these three bullet points should begin with lowercased words and have consistent EOL punctuation

File: CONTRIBUTING.md
Comment (jonatack) [inline CONTRIBUTING.md:92]: - "other global" could be "other transversal"
- can omit "that touch...as non-GUI code,"
- s/the ... be used./should be opened in bitcoin/bitcoin./

File: CONTRIBUTING.md
Comment (jonatack) [inline CONTRIBUTING.md:88]: for clarity and to avoid centralisation, would suggest avoiding the term "main" in favor of bitcoin-core/gui and bitcoin/bitcoin

File: CONTRIBUTING.md
Comment (jonatack) [inline CONTRIBUTING.md:98]: suggestions:
- use active rather than passive voice
- begin by stating in which case this applies

File: README.md
Comment (jonatack) [inline README.md:28]: missing verb: s/but is/but it is/ (or omit the comma)

File: CONTRIBUTING.md
Comment (jonatack) [inline CONTRIBUTING.md:87]: I think it may be worthwhile to state here from the start that the bitcoin-core/gui repo master branch is identical to the master branch in bitcoin/bitcoin. This may avoid some confusion while reading the rest of the section.

**Review by jonatack — COMMENTED** [review]
Some suggested changes below.

File: CONTRIBUTING.md
Comment (MarcoFalke) [reply CONTRIBUTING.md:88]: "main repository" seems like a good way to abbreviate "bitcoin/bitcoin repository" or "non-gui repository".

Switched to "core repository". Hope that's ok as well

**Comment by MarcoFalke** [comment]
Addressed @jonatack's feedback

**Comment by practicalswift** [comment]
ACK fae0fe0f6e7bee941dd107241d378081f68945fc

**Review by jonatack — COMMENTED** [review]
ACK modulo terminology nit s/"node"/"bitcoin/bitcoin"/. I am unsure if separating the repos will be beneficial but (a) I know there is much that I don't know, and (b) I empathise with the sentiments and logic expressed by the maintainers and long-term contributors in favor of doing so, or at least trying it.

**Comment by troygiorshev** [comment]
ACK, with the following nit
In https://raw.githubusercontent.com/bitcoin/bitcoin/master/.github/ISSUE_TEMPLATE.md#L12
`<!--- What was the actual behavior (provide screenshots if the issue is GUI-related)? -->`

a removal of comment in brackets and an addition of something like:
```
GUI-related issues should be opened against
https://github.com/bitcoin-core/gui
first. See CONTRIBUTING.md
```

**Comment by MarcoFalke** [comment]
Added note as requested by @troygiorshev 

> ACK modulo terminology nit s/"node"/"bitcoin/bitcoin"/

Is there an issue with "node"? To me it seems an appropriate choice to abbreviate `bitcoin/bitcoin`, so I left it as is for now.

**Comment by fjahr** [comment]
ACK faceed753a4d3d909985cdfc42b23f5dd395e168

I am curious to see how this plays out. Generally, I share @jonatack 's concerns but unless the major GUI contributors are against it (which does not seem to be the case), it is certainly worth a try. Are there already thoughts on how and when this should be evaluated to decide if this is a success or a failure?

**Comment by MarcoFalke** [comment]
I was planning to open a re-evaluation issue to collect concerns and feedback over time. If the negative feedback outweighs the positive feedback, it should be reverted without having to wait for an explicit re-evaluation point.

Though, I am also happy to host a topic in the IRC meeting in 6 months and 12 months to re-evaluate the move and consider re-adjustments.

**Comment by Sjors** [comment]
ACK 66666d55b16603287d30ee672061ced8fc6752ab

**Review by hebasto — APPROVED** [review]
re-ACK 66666d55b16603287d30ee672061ced8fc6752ab

**Comment by promag** [comment]
ACK

**Comment by troygiorshev** [comment]
re-ACK 66666d5

**Comment by practicalswift** [comment]
re-ACK 66666d55b16603287d30ee672061ced8fc6752ab

**Comment by MarcoFalke** [comment]
Ok, going to merge this. Let's continue discussion in https://github.com/bitcoin-core/gui/issues/5

**Comment by ErikThiart** [comment]
Wow, awesome.

**Comment by JeremyCrookshank** [comment]
ACK

**Pull request merged**
